src/journal/journald-file.c

```diff
@@ -46,6 +51,25 @@ static void journald_file_set_offline_internal(JournaldFile *f) {
 
                         f->file->header->state = f->file->archive ? STATE_ARCHIVED : STATE_OFFLINE;
                         (void) fsync(f->file->fd);
+
+                        /* If we're on BTRFS, let's rewrite the journal file with COW enabled when archiving.
+                         * We're finished writing the file so COW won't impact write performance anymore and
+                         * COW provides many benefits such as checksumming and compression. */
+
+                        if (f->file->archive) {
+                                r = fd_is_fs_type(f->file->fd, BTRFS_SUPER_MAGIC);
+                                if (r < 0)
+                                        log_debug_errno(r, "Failed to determine if journal is on btrfs: %m");
+                                if (r <= 0)
+                                        continue;
```

Comment:
hmm, instead of doing an explicit check for btrfs, maybe try to turn off FS_NOCOW_FL explicitly. If that fails, then do the copy, but if it works just leave it as is. After al FS_NOCOW_FL is defined generically, even if btrfs right now is the primary (only?) user for it. Moreover, if I read the upstream discussions above, even btrfs might possible allow FS_NOCOW_FL to be turned off after the file is populated one day.

chattr_full() already has an argument that returns which flags were actually set. And it should be really simple to check if FS_NOCOW_FL is among them, and if not, just do the copy dance. With that in place things will work great on btrfs without hardcoding anything to btrfs behaviour here.

Reply:
Done, although I had to make chattr_full() extra paranoid since btrfs seems to silently discard the FS_NOCOW_FL flag if there's data in the file already, trying to disable it doesn't even return an error.